### PR TITLE
Board theme

### PR DESCRIPTION
**Pull request by MathieuDreano**
Allows for different board backgrounds options
Maybe create a new branch from this for further contribution before production
possible improvement:
- custom background (from url)
- custom stones color

### CODE REVIEW

**Comment by acristescu** [comment]
Hi @MathieuDreano thank you for your contribution. I will make a proper review the next few days, but after a 5-minute look I had a few requests:
- what does HNG stand for and why would people need a bright aqua background? Does it have something to do with accessibility?
- the app/src/main/res/texture/mipmap-night-xxxhdpi/hng.jpg file is wrong I think, it's a copy of the wood theme under a different name
- the themes we have so far seem to be solid colours. Could we treat them as such by either just drawing a rect with the appropriate background colour OR alternatively having an android vector drawable instead? Jpegs are really bad for APK size and also for performance.

**Comment by MathieuDreano** [comment]
Hi Alexandru,
- I have no idea what HNG stands for, i was just copying some of OGS theme options. (this is actually the theme I use myself, maybe a color picker would give more options, I'll dig into this)
image
- I removed app/src/main/res/texture/mipmap-night-xxxhdpi/hng.jpg as it was unused.
- good idea for drawable, I'll try to fix that and use jpg only when needed

File: app/src/main/java/io/zenandroid/onlinego/ui/screens/settings/CustomAdapter.kt
Comment (acristescu) [inline app/src/main/java/io/zenandroid/onlinego/ui/screens/settings/CustomAdapter.kt:50]: I think what you're looking for is ContextCompat.getColor(): https://stackoverflow.com/questions/31590714/getcolorint-id-deprecated-on-android-6-0-marshmallow-api-23

File: app/src/main/java/io/zenandroid/onlinego/ui/screens/settings/CustomAdapter.kt
Comment (acristescu) [inline app/src/main/java/io/zenandroid/onlinego/ui/screens/settings/CustomAdapter.kt:41]: In kotlin we like to avoid double bang (!!) as much as possible as it circumvents the platform's nullability checks. Please try to use ?. or ?.let { someCode() } wherever possible.

File: app/src/main/java/io/zenandroid/onlinego/ui/screens/settings/CustomAdapter.kt
Comment (acristescu) [inline app/src/main/java/io/zenandroid/onlinego/ui/screens/settings/CustomAdapter.kt:16]: Sometimes I forget how impossibly complicated the legacy view system is... At some point I'll have to reimplement the settings in Compose as displaying a list should not be that complicated...

Comment (MathieuDreano) [reply app/src/main/java/io/zenandroid/onlinego/ui/screens/settings/CustomAdapter.kt:16]: And I didn't even manage to put a grid layout on this dialog.... It is hell for me 🔥

File: app/src/main/java/io/zenandroid/onlinego/ui/screens/settings/CustomAdapter.kt
Comment (MathieuDreano) [reply app/src/main/java/io/zenandroid/onlinego/ui/screens/settings/CustomAdapter.kt:41]: Ok, never used Kotlin before, I'll look into it to avoid breaking your work

**Comment by MathieuDreano** [comment]
Made required changes, let me know if you see anything else.

File: app/src/main/java/io/zenandroid/onlinego/data/model/BoardTheme.kt
Comment (acristescu) [inline app/src/main/java/io/zenandroid/onlinego/data/model/BoardTheme.kt:68]: woah, semi-colon... that brings back memories of my Java days, but it's completely unnecessary in Kotlin

File: app/src/main/java/io/zenandroid/onlinego/ui/composables/BoardComposable.kt
Comment (acristescu) [inline app/src/main/java/io/zenandroid/onlinego/ui/composables/BoardComposable.kt:444]: typo: tex instead of text

File: app/src/main/java/io/zenandroid/onlinego/ui/composables/BoardComposable.kt
Comment (acristescu) [inline app/src/main/java/io/zenandroid/onlinego/ui/composables/BoardComposable.kt:446]: isn't this done?

**Review by acristescu — CHANGES_REQUESTED** [review]
Hi @MathieuDreano sorry for the horrible delay, life is pretty crazy right now. There are a few minor fixes that need to be done before we can merge this if you're still interested in contributing.

File: app/src/main/java/io/zenandroid/onlinego/ui/composables/BoardComposable.kt
Comment (MathieuDreano) [reply app/src/main/java/io/zenandroid/onlinego/ui/composables/BoardComposable.kt]: I had trouble because the MyGameUI would not update the boards when the setting changed. We had no problem with coordinates as they are never shown here whatever the setting is.
You'll see that I made a change in the onViewResumed() of MyGamesViewModel, there is certainly a much more beautiful way to achieve this, but I'm not confident with the update lifecycle of this part...

**Comment by MathieuDreano** [comment]
I had trouble because the MyGameUI would not update the boards when the setting changed.
We had no problem with "showcoordinates" as they are never shown in this view whatever the setting is.

You will see that I made a change in the onViewResumed() of MyGamesViewModel, there is certainly a much more beautiful way to achieve this, but I'm not confident with the update lifecycle of this part...

**Comment by acristescu** [comment]
> I had trouble because the MyGameUI would not update the boards when the setting changed. We had no problem with "showcoordinates" as they are never shown in this view whatever the setting is.
> 
> You will see that I made a change in the onViewResumed() of MyGamesViewModel, there is certainly a much more beautiful way to achieve this, but I'm not confident with the update lifecycle of this part...

Heh, welcome to Android development, when in doubt, we just dump stuff into `onResume()` :D I think this will be fine, the only downside is that SharedPreferences.get() does actually do I/O (it opens a file and parses it) so it's a bit slow which could cause problem on very slow phones, as you're not supposed to block the main thread with I/O. We could either:
1.  cache the current theme in memory (at the repository level)
2. use a coroutine to subscribe to read (needlessly complicated).
3. migrate to the new "Preferences DataStore" library that I haven't had time to look at.
4. use a global SnapshotState (e.g. `by mutableStateOf()` as in the GameViewModel) to remember the BoardTheme. The advantage of a SnapshotState is that if the value changes all the Composables that read it get invalidated causing the value to be read again automagically. Question is on which object we put that global SnapshotState and when do we initialize it with the saved value.

But mostly it's fine as it is, I was just mentioning these solutions in case you want to investigate further ways to fix it.

I think we could even merge this, it depends if you want to do further work on this. Here are a few minor points that could be worked on (perhaps in a later PR?):

- the name of the theme is displayed as "HNG_DARK" in the settings, which is not nice, the enum could have a pretty name string that we show to the user without all caps and underscores etc.
- you mentioned you want to implement custom stones
- some of the text colors are rather hard to read, such as the grey on WOOD_DARK

**Comment by MathieuDreano** [comment]
I think it's already a good start, i modified to add some more meaningful theme name.
Custom stones and custom background from URL could be in a different PR, one day....

**Review by acristescu — APPROVED** [review]

**Comment by benjaminpjones** [comment]
I'm random GitHub creeping... 😋

> what does HNG stand for and why would people need a bright aqua background?

> I have no idea what HNG stands for, i was just copying some of OGS theme options.

I gather this stands for Hikaru No Go (though I don't know anything about the series or why it is bright blue).

**Comment by acristescu** [comment]
I wouldn't be surprised if that is actually an error on the site and we just brought it in the app because it "looks cool" :D

**Comment by mdreano** [comment]
It's already been renamed to CYAN in the code ;)
@benjaminpjones yep, came to the same conclusion HikaruNoGo, and like you i didn't find anything related to a cyan board 😄

**Pull request merged**
